[wook] step15-1 리뷰 부탁드립니다!

pocokim opened the pull request:
이벤트 리스너를 컨트롤러에 두었습니다. 역시나 컨트롤러는 복잡해졌습니다만 
model , view는 상대적으로 깔끔한것같습니다.

컨트롤러에서 반복이 발생하는 부분이 있습니다만 이것을 함수로 분리하기는 살짝 애매한것같습니다.
예를 들어 index를 바꾸고, render를 하고, css효과를 주는 함수들을 실행하는것인데요. 
단순 반복을 방지하기위해 
여러 함수를 실행하는 하나의 함수를 만드는것이 그렇게 좋은 방법인지 모르겠어 일단 두었습니다. 

[데모](https://dxvinci.github.io/codesquad/step15/index.html)입니다.  감사합니다 🙇‍♂️

pocokim committed:
feat: 검색기능 마크업 구현

pocokim committed:
refactor: 디렉토리 구조 변경

캐러샐관련 모듈은 캐러샐 폴더에 존재하도록 구조 변경

pocokim committed:
feat: 자동완성 mvp 구현

input 이벤트가 발생시 value를 searchModel에 보내고 해당하는 데이터를 matchedView에 출력하도록 구현하였다.

이벤트등록을 controller에서 하고
server 데이터를 가져오는것을 model에서 하는것이 적합한지는 계속 리팩토링 할 예정

pocokim committed:
chore: 자동완성 View 마크업 수정

1. header 태그 생성하여 뒷배경추가하고
2. position relative 가 잘못결려서 ul태그의 위치를 수정함

pocokim committed:
feat: 자동완성시 화면 렌더 mvp

pocokim committed:
feat: sleep 함수 구현

pocokim committed:
feat: 자동완성시 mvp2 구현

li template 가 계속 추가되지 않고 없으면만들고, 있으면 교체하는 로직으로 수정

pocokim committed:
refactor: sleep 함수를 모듈로 분리

pocokim committed:
refactor: render 함수

fast-return 으로  있을때 삭제하도록하고,

반복되는 템플릿추가와 insert 로직을 조건문에서 분리

pocokim committed:
feat: 키보드위아래 mvp

inputViewKeyDownHandler는
curIdx 와 prevIdx를 받아
index에 맞는 li의 배경색을 바꾸고 idx를 변경하도록 하였다.
inputViewKeyDownHandler안에서
두가지 역할이 동시에 일어나므로
분리가 필요하며
위에서 아래로 바뀔때 버그가있다.

pocokim committed:
feat: 키보드위아래 이동 하이라이트구현

현재index로 그리고 ++를 하고, --를 하는 방식으로 구현하면 좌우변경시 에러가발생한다.
따라서 그리고 연산하는것이 아닌,

<연산이 끝난다음에 그리도록 해야한다>
설명하기는 애매하지만, 연산이 끝난결과를 그리는것이 훨씬 더 직관적이다.

왜냐면 아래방향키를 눌렀을때 그 연산후 그리고, 그리고 위방향키를 눌렀을때 또 그것에따라 연산한 결과를 그리기 때문이다.

pocokim committed:
키보드 이동 무한 구현

pocokim committed:
하이라이트 기능 mvp 구현

replace를 이용해 새로운 태그를 추가함.

그런데 이때 bold를 css로 제어해야하는 이유를 모르겠음. 그냥 태그만 추가해도 원래는 bold가 추가되어야하지 않나?

pocokim committed:
chore: 키보드이동시 배경색 변경

important! 를 써야하는 이유를 모르겠음.

pocokim committed:
선택한 상태에서 엔터키 입력시 inputView 추가기능 구현

의사코드를 먼저 만들고, 그떄 필요한 함수를
먼저 작명하고 클래스에 들어가서 만드는 방식으로 코드를 작성하였다.
매우잘된 사례여서 커밋에 기록!!!

pocokim committed:
feat : 최근검색내역 mvp 구현

1. historyView의 render 함수 구현
2. searchModel의 save 함수 구현
3. App과 Controller에 historyView 추가

pocokim committed:
chore: 매치된 단어가 없을때 검색내역숨기기

1. render함수에서 deconstructing 위치를 조절하여 해결하였음!

논리상으로도 undefined가 위로 가는게 더 올바름!

2. matchedView가 있을때는 matchedView를 hide 하는것이 아니라 historyView로 변경해서 render해야하며 이것을 추가적으로 구현해야함.

pocokim committed:
feat: 최근검색 5개 역순 노출 구현

1. HistoryView의 render함수에서 immutable 방식으로 historys 배열을 제어하려고 했으나 장점을 모르겠어서 이정도에서 끝냄.

2. render 함수에서 뒤집고, 다섯개까지만 노출되는것을 코드로 알아보기가 살짝 어려울것같다는 생각이 들음... 어떻게 처리하는게 좋을지? 명시적으로 알기위해 2줄짜리라도 함수로 분리하고 render에서는 선언하는식으로 함수를 짜는게 떠오르는데, 읽기는 좋지만 그렇게 아주 미세한것도 함수로 나누는게 좋을지는 의문 (상속의 관점에서는 좋을것같다는 생각을 해봄)

pocokim committed:
chore: 새로 검색시 index 초기화

 this.matchedView.curserIndex = -1; 를 명시적으로 보여주기위해 함수로 만드는게 좋을지는 의문

pocokim committed:
feat: 버튼, 자동완성li 클릭시 저장기능 구현

mousedown 전에 mouseover 됬을때 커서 잡히는 효과를 주는게 자연스러운듯 하며

반복되는 기능이 많아서 함수로 분리하고자하는데 어떻게 분리하는게 좋을지 의문.. 리팩토링이 필요하다.

pocokim committed:
chore: 불필요한 주석 삭제

crongro commented on Search/SearchModel.js:
```diff
@@ -0,0 +1,19 @@
+  }
+
+  async find(input) {
+    const dataURL = `https://h3rb9c0ugl.execute-api.ap-northeast-2.amazonaws.com/develop/amazon_autocomplete?query=${input}`;
```
이런 URL은 개발서버/리얼서버마다 다른게 일반적이니 searchModel.js 밖에서 선언하는 게 좋죠.

crongro commented on Search/SearchModel.js:
```diff
@@ -0,0 +1,19 @@
+    const dataURL = `https://h3rb9c0ugl.execute-api.ap-northeast-2.amazonaws.com/develop/amazon_autocomplete?query=${input}`;
+    const response = await fetch(dataURL);
+    const fetchedData = await response.json();
+    if(fetchedData.statusCode === 404) return; //데이터가 없을때 예외처리
```
try - catch 를 써서 async 에서 생기는 오류를 처리해보세요.

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,123 @@
+    await this.matchedView.render(matchedData, inputValue);
+  }
+
+  inputViewKeyDownHandler(e) {
```
몇 개 문제가 있죠.  개선하면 더 좋을거 같아요.
- 너무 크다.
- if안에 if문이 존재한다
- arrowdown/arrowup 간에 중복이 보인다.
- arrowdown/arrowup 안의 코드를 하위함수로 분리하는 게 좋겠다.(위에 중복도 같이 해결하면서)

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,123 @@
+    });
+
+    this.matchedView.ul.addEventListener("mouseover", e => {
+      const fetchedValue = e.target.innerText;
```
몇번 강조한 destructuring을 잘 활용해서 표현을 간결하게 해보세요.

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,123 @@
+    this.registerEvents();
+  }
+
+  registerEvents() {
```
이부분의 이벤트리스너를 모두 분리해서 선언하면  
'어떤 이벤트를 등록하는지'를 쉽게 알 수 있을 거 같아요.

crongro commented on Search/SearchModel.js:
```diff
@@ -0,0 +1,19 @@
+    this.historyQueue = [];
+  }
+
+  async find(input) {
```
find보다는 결국 get하는거니까, getData라고 하는 식이  더 낫죠.

crongro commented on Search/sleep.js:
```diff
@@ -0,0 +1,3 @@
+export default function sleep(ms) {
```
sleep.js 하나를 덩그러니 별도 파일로 두지 말고, util이나 library 와 같은 파일을 만들고 추가하는 게 좋겠죠.

pocokim replied:
sleep 함수는 util로 만들기에 함수가 하나밖에 없는것같아서.. 의도적으로 만들지 않았는데요 ㅎㅎ... 그래도 만드는게 좋을까요? ㅎㅎ

crongro commented on Search/InputView.js:
```diff
@@ -0,0 +1,12 @@
+  }
+
+  render(value){
+    this.div.value = value;
```
div 가 아닌거 같은데, input태그 아닌가요?

crongro commented on Search/HistoryView.js:
```diff
@@ -0,0 +1,32 @@
+
+    this.initializeTemplate();
+    let historys = data.reverse();
+    historys = historys.slice(0, 5); //  immutable 연습하기 splice나 spread operator를 이용해서 만들 수 있지만 이렇게 해도 되지않나?
```
새로운 배열을 만드는 것이면 immutable 맞겠죠.
생각한 것처럼, immutable 방식은 다양할 수 있어요.

crongro commented on Search/HistoryView.js:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  initializeTemplate() {
+    if (this.liTemplates) {
```
if(!this.liTemplates) return;

crongro reviewed (changes requested):
리뷰가 좀 늦었네요;

module을 잘 나눠서 개발했지만,  수정할 게 좀 보여요!

pocokim committed:
refactor: URL , try ~ catch

pocokim committed:
refactor: try ~ catch 개선

404일때 데이터가 없는것이 정상이라고 판단되어 catch에서 처리하지 않음.

나머지 알수없는(?) 에러만 catch에 넣음.

pocokim committed:
chore : fast- return , 함수와 변수명 변경

pocokim committed:
refactor: inputViewKeyDownHandler

1. fast-return 으로 뎁스 낮춤
2. changeIdx 함수에서 3항연산자를 이용해 조건에 들어갈 변수를 수정
3. arrowUp , down 일때 동시에 처리

pocokim committed:
chore : deconstructing

pocokim committed:
refactor: delete controller.init()

pocokim committed:
refactor: inputView에서 이벤트등록하도록 변경

1. input, keyDown , click 이벤트 등록
2. clickHandler 함수에 바인딩하기 위해 click이벤트를 처리하던 코드를 inputViewClickHandler라는 함수로 분리 (근데 이 함수 어디에 놔야 가독성이 좋을지 고민)

inputViewHandler 함수안에서 call처럼 this를 넘겨주어야하는건 없을까? 고민하였지만 일단 잘 되고.. 안될때 디버깅하는걸로!!

pocokim committed:
refactor: matchedView에 이벤트 등록

matchedView의 handler 함수는 arrow function으로 바인딩없이 사용하도록 함.

pocokim committed:
chore: delete magic number

pocokim committed:
chore: 함수명 변경

1. find -> get으로
2. changeCurrIdx를 changeMatchedViewIdxInfinite

pocokim committed:
refactor: 404 인경우 catch 에서 수정하도록 변경

crongro commented on Search/InputView.js:
```diff
@@ -1,12 +1,28 @@
+
+    this.input.addEventListener("keydown", e => this.keyDownHandler(e));
+
+    this.btn.addEventListener("click", e => {
```
arrow function을 썼지만 어디는 한줄 어디는 세줄.. 맞추면 더 이쁠 듯

crongro commented on Search/SearchController.js:
```diff
@@ -102,17 +48,61 @@ class SearchController {
+    this.searchModel.save(fetchedValue);
+    this.inputView.render(fetchedValue);
+    this.matchedView.hide();
+    this.matchedView.curserIndex = -1;
```
별건 아니지만 this.matchedView.setCurserIndex(-1);
이렇게 메서드를 통해서 접근하는게 더 명시적인 방법이죠.
속성을 직접 접근하는 건 좋지 않아요. 
메서드가 일종의 인터페이스라고 생각하세요.

crongro reviewed (approved):
리팩토링을 이정도 수준으로만 해도 좋은거 같아요.
잘 수정했어요.
